File: src/main/java/net/cjsah/console/command/builder/RequiredArgumentBuilder.java
package net.cjsah.console.command.builder;

import net.cjsah.console.command.argument.Argument;
import net.cjsah.console.command.tree.ArgumentCommandNode;
import net.cjsah.console.command.tree.CommandNode;

public class RequiredArgumentBuilder<T> extends ArgumentBuilder<RequiredArgumentBuilder<T>>{
    private final String name;
    private final Argument<T> type;

    private RequiredArgumentBuilder(final String name, final Argument<T> type) {
        this.name = name;
        this.type = type;
    }

    public static <S> RequiredArgumentBuilder<S> argument(String name, Argument<S> type) {
        return new RequiredArgumentBuilder<>(name, type);
    }

    @Override
    protected RequiredArgumentBuilder<T> getThis() {
        return this;
    }

    @Override
    public CommandNode build() {
        CommandNode result = new ArgumentCommandNode<>(this.getHelp(), this.name, this.type, getCommand(), getRequirement());
        getArguments().forEach(result::addChild);
        return result;
    }
}

File: src/main/kotlin/net/cjsah/console/command/builder/RequiredArgumentBuilder.kt
package net.cjsah.console.command.builder

import net.cjsah.console.command.argument.Argument
import net.cjsah.console.command.tree.ArgumentCommandNode

class RequiredArgumentBuilder<T> private constructor(
    private val name: String,
    private val type: Argument<T>
): ArgumentBuilder<RequiredArgumentBuilder<T>>() {

    companion object {
        @JvmStatic
        fun <S> argument(name: String, type: Argument<S>) = RequiredArgumentBuilder(name, type)
    }


    override fun getThis() = this

    override fun build() = ArgumentCommandNode(getHelp(), name, type, getCommand(), getRequirement()).apply {
        getArguments().forEach { this.addChild(it) }
    }

}